src/frontend/flatten/yaksa_unflatten.c

```diff
@@ -9,38 +9,35 @@
 #include <string.h>
 #include <stdlib.h>
 
-static int unflatten(yaksi_type_s ** type, const void *flattened_type)
+static inline int unflatten(yaksi_type_s ** type, const void *flattened_type)
 {
     int rc = YAKSA_SUCCESS;
     yaksi_type_s *newtype;
     const char *flatbuf = (const char *) flattened_type;
 
-    rc = yaksi_type_alloc(&newtype);
-    YAKSU_ERR_CHECK(rc, fn_fail);
-
-    /* don't overwrite the local ID, but keep the original ID that we
-     * got over the flattened type */
-    unsigned int local_id, orig_id;
-    local_id = newtype->id;
-    memcpy(newtype, flatbuf, sizeof(yaksi_type_s));
-    flatbuf += sizeof(yaksi_type_s);
-    orig_id = newtype->id;
-    newtype->id = local_id;
-    yaksu_atomic_store(&newtype->refcount, 1);
+    if (((yaksi_type_s *) flattened_type)->kind == YAKSI_TYPE_KIND__BUILTIN) {
+        int id = ((yaksi_type_s *) flattened_type)->id;
+        yaksi_type_s *tmp;
+        rc = yaksi_type_get(id, &tmp);
+        YAKSU_ERR_CHECK(rc, fn_fail);
+        yaksu_atomic_incr(&tmp->refcount);
+        newtype = tmp;
+        goto setup_hooks;
+    } else {
+        rc = yaksi_type_alloc(&newtype);
+        YAKSU_ERR_CHECK(rc, fn_fail);
+
+        /* don't overwrite the local ID, but keep the original ID that
+         * we got over the flattened type */
```

Comment:
Is this comment for the builtin case?

Reply:
Part of it.  The "don't overwrite the local ID" part should be there.  I'll redo the comment when I repush.  Is that the only comment?